### PR TITLE
Add eventbus to cloudwatch event rule

### PR DESCRIPTION
**Pull request by gdavison**
Completes work from #12886 to add custom event buses to CloudWatch Events rules.

Closes #12886
Relates #9330 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* resource/aws_cloudwatch_event_rule: Add `event_bus_name`
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCloudWatchEventRule_'

--- PASS: TestAccAWSCloudWatchEventRule_Name_Generated (119.30s)
--- PASS: TestAccAWSCloudWatchEventRule_NamePrefix (124.29s)
--- PASS: TestAccAWSCloudWatchEventRule_ScheduleAndPattern (124.30s)
--- PASS: TestAccAWSCloudWatchEventRule_role (130.35s)
--- PASS: TestAccAWSCloudWatchEventRule_description (165.74s)
--- PASS: TestAccAWSCloudWatchEventRule_pattern (167.54s)
--- PASS: TestAccAWSCloudWatchEventRule_IsEnabled (198.77s)
--- PASS: TestAccAWSCloudWatchEventRule_basic (198.91s)
--- PASS: TestAccAWSCloudWatchEventRule_EventBusName (201.43s)
--- PASS: TestAccAWSCloudWatchEventRule_tags (223.86s)
```

### CODE REVIEW

**Comment by awendt** [comment]
@gdavison Thanks for working on this. We created a fork based on 3.12.0, merged in your changes from this PR (and `resource_aws_cloudwatch_event_target`) but noticed we cannot import rules that are defined on an event bus different from `default`. Here's a minimal config:

```hcl
provider "aws" {
  version = "3.12.0-with-15727-and-15799"

  region = "eu-west-1"
}

data "aws_caller_identity" "current" {}

resource "aws_cloudwatch_event_bus" "test" {
  name = "test-awendt"
}

resource "aws_cloudwatch_event_rule" "non-default-bus" {
  event_bus_name = aws_cloudwatch_event_bus.test.name

  description    = "Filters all events from EventBridge Bus ${aws_cloudwatch_event_bus.test.name}"
  is_enabled     = false
  name           = "test-rule"
  event_pattern  = jsonencode(
    {
      account = [
        data.aws_caller_identity.current.account_id,
      ]
    }
  )
}
```

## Steps to reproduce

1. Create the event bus using AWS CLI: `aws events create-event-bus --name test-awendt`
2. Create an event rule on that bus: `aws events put-rule --name test-rule --state DISABLED --event-pattern "{\"account\": [\"AWS_ACCOUNT_ID\"]}" --event-bus-name test-awendt` (you need to interpolate the account ID)
3. Import the event bus: `terraform import aws_cloudwatch_event_bus.test test-awendt`
4. Import the event rule: `terraform import aws_cloudwatch_event_rule.non-default-bus test-rule`

This last step fails with:

```
aws_cloudwatch_event_rule.non-default-bus: Importing from ID "test-rule"...
aws_cloudwatch_event_rule.non-default-bus: Import prepared!
  Prepared aws_cloudwatch_event_rule for import
aws_cloudwatch_event_rule.non-default-bus: Refreshing state... [id=test-rule]

Error: Cannot import non-existent remote object

While attempting to import an existing object to
aws_cloudwatch_event_rule.non-default-bus, the provider detected that no
object exists with the given id. Only pre-existing objects can be imported;
check that the id is correct and that it is associated with the provider's
configured region or endpoint, or use "terraform apply" to create a new remote
object for this resource.
```

Can you reproduce this? What could be causing this behavior?

**Comment by awendt** [comment]
@gdavison Some more context: We noticed Terraform wanted to create a resource that was already existing. It seems both `refresh` and `import` are not working as intended.

**Comment by gdavison** [comment]
Hi @awendt. When importing a rule for a non-default bus, the format of the import id is <bus-name>/<rule-name>, and it assumes that the bus is the default bus if only the <rule-name> is passed. In the example you posted, you'd need to call `terraform import aws_cloudwatch_event_rule.non-default-bus test-awendt/test-rule`.

**Review by bflad — APPROVED** [review]
One small question, otherwise LGTM 🚀 

Output from acceptance testing:

```
--- PASS: TestAccAWSCloudWatchEventRule_ScheduleAndPattern (34.58s)
--- PASS: TestAccAWSCloudWatchEventRule_Name_Generated (38.15s)
--- PASS: TestAccAWSCloudWatchEventRule_NamePrefix (39.13s)
--- PASS: TestAccAWSCloudWatchEventRule_role (53.03s)
--- PASS: TestAccAWSCloudWatchEventRule_pattern (56.78s)
--- PASS: TestAccAWSCloudWatchEventRule_description (58.19s)
--- PASS: TestAccAWSCloudWatchEventRule_IsEnabled (68.64s)
--- PASS: TestAccAWSCloudWatchEventRule_basic (71.60s)
--- PASS: TestAccAWSCloudWatchEventRule_EventBusName (72.20s)
--- PASS: TestAccAWSCloudWatchEventRule_tags (82.68s)
```

**Comment by awendt** [comment]
@gdavison Thanks for catching this, I can confirm the import works as you described. 🎉 

I was also able to track down the root cause why refresh isn't working for us, and I believe we can just re-import the resources and all will be good.

File: aws/resource_aws_cloudwatch_event_rule.go [hunk 1]
```diff
@@ -263,25 +281,31 @@ func buildPutRuleInputStruct(d *schema.ResourceData, name string) (*events.PutRu
...
-	}
-
-	if v, ok := d.GetOk("tags"); ok {
-		input.Tags = keyvaluetags.New(v.(map[string]interface{})).IgnoreAws().CloudwatcheventsTags()
+		scheduleExpression := v.(string)
+		if eventBusName == tfevents.DefaultEventBusName {
+			input.ScheduleExpression = aws.String(scheduleExpression)
+		} else {
+			return nil, errors.New("schedule_expression can only be set on the default event bus")
+		}
```
[the remark this answers is not in the document]
Comment (gdavison) [reply aws/resource_aws_cloudwatch_event_rule.go:308]: This is short-circuiting the API call. I would have set it as a `ValidateFunc` if they supported cross-attribute validation

**Pull request merged**